### PR TITLE
Implementing user functionality

### PR DESCRIPTION
**Pull request by arnaudoff**

### CODE REVIEW

File: app/api/User.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+case class UserResponse(id: Long, name: String)
+object UserResponse {
+  def fromDb(user: db.User): UserResponse = UserResponse(user.id, user.name)
+  implicit lazy val jsFmt: Format[UserResponse] = Json.format[UserResponse]
+}
+
+sealed trait CreateUserResponse
+object CreateUserResponse {
+  case class Created(user: UserResponse) extends CreateUserResponse
+  case class NotCreated(message: String) extends CreateUserResponse
```
Comment (PlamenNeshkov) [inline app/api/User.scala]: `error` would be more precise

File: app/api/User.scala [hunk 2]
```diff
@@ -0,0 +1,36 @@
+package api
+
+import play.api.libs.json._
+
+case class UserResponse(id: Long, name: String)
```
Comment (PlamenNeshkov) [inline app/api/User.scala]: Should be called `UserRetrieveResponse` and be a union type, analogous to `StopRetrieveResponse`

File: app/api/User.scala [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+  val createdJsFormat: Format[Created] = Json.format[Created]
+  val notCreatedJsFormat: Format[NotCreated] = Json.format[NotCreated]
+  implicit lazy val jsWrites: Writes[CreateUserResponse] = Writes({
+    case c: Created => createdJsFormat.writes(c)
+    case nc: NotCreated => notCreatedJsFormat.writes(nc)
+  })
+}
+
+case class UserRequest(name: String, password: String)
```
Comment (PlamenNeshkov) [inline app/api/User.scala]: `UserCreateRequest`

File: app/api/User.scala [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+import play.api.libs.json._
+
+case class UserResponse(id: Long, name: String)
+object UserResponse {
+  def fromDb(user: db.User): UserResponse = UserResponse(user.id, user.name)
+  implicit lazy val jsFmt: Format[UserResponse] = Json.format[UserResponse]
+}
+
+sealed trait CreateUserResponse
```
Comment (PlamenNeshkov) [inline app/api/User.scala]: `UserCreateResponse`

File: app/api/User.scala [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    case c: Created => createdJsFormat.writes(c)
+    case nc: NotCreated => notCreatedJsFormat.writes(nc)
+  })
+}
+
+case class UserRequest(name: String, password: String)
+object UserRequest {
+  implicit lazy val userRequestReads = Json.reads[UserRequest]
+
+  def fromJsValue(value: JsValue): Option[UserRequest] = {
```
Comment (PlamenNeshkov) [inline app/api/User.scala]: Why is this needed? Use `play.api.json`

File: app/controllers/UsersController.scala [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class UsersController @Inject()(cc: ControllerComponents, usersRepo: db.UsersRepo)(implicit ec: ExecutionContext)
+  extends AbstractController(cc) {
+
+  private[this] def okJson[T](t: T)(implicit writes: Writes[T]) = Ok(Json.toJson(t))
+
+  def create = Action.async { implicit request =>
+    val userRequest = api.UserRequest.fromJsValue(request.body.asJson.get)
```
Comment (PlamenNeshkov) [inline app/controllers/UsersController.scala]: Don't do this, take a look at [this](https://playframework.com/documentation/2.0/ScalaJsonRequests)

```
Action.async(parse.json) { request =>
    request.body.as[Type]
    ...
}
```

Also, I'm pretty sure `request` does not need to be `implicit`

File: app/db/UsersRepo.scala [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+import javax.inject.{Inject, Singleton}
+import play.api.db.slick.DatabaseConfigProvider
+
+import scala.concurrent.{ExecutionContext, Future}
+
+case class User(id: Long, name: String, password: String)
+case class FavouriteStop(id: Long, userId: Long, stopId: String)
+
+@Singleton
+class UsersRepo @Inject()(protected val dbConfigProvider: DatabaseConfigProvider)(implicit ec: ExecutionContext)
```
Comment (PlamenNeshkov) [inline app/db/UsersRepo.scala]: Adhere to the existing indentation/newline style for `Repo`s

File: app/db/UsersRepo.scala [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    def id = column[Long]("id", O.PrimaryKey, O.AutoInc)
+    def name = column[String]("name")
+    def password = column[String]("password")
+    def * = (id, name, password) <> ((User.apply _).tupled, User.unapply)
+  }
+
+  protected val usersQuery = TableQuery[UserTable]
+}
+
+private[db] trait FavouriteStopTable extends UserTable with StopTable { this: SlickComponent =>
```
Comment (PlamenNeshkov) [inline app/db/UsersRepo.scala]: `protected[db]`, we want to access it in other files in this package in the future

File: app/db/UsersRepo.scala [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+  def addFavouriteStop(userId: Long, stopId: String): Future[FavouriteStop] = db.run {
+    (favouriteStopsQuery.map(fs => (fs.userId, fs.stopId))
+      returning favouriteStopsQuery.map(_.id)
+      into ((userStopIds, id) => FavouriteStop(id, userStopIds._1, userStopIds._2))
+      ) += (userId, stopId)
+  }
+}
+
+private[db] trait UserTable { this: SlickComponent =>
```
Comment (PlamenNeshkov) [inline app/db/UsersRepo.scala]: Same, `protected[db]`

File: app/db/UsersRepo.scala [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+  private[FavouriteStopTable] class FavouriteStopTable(tag: Tag)
+    extends Table[FavouriteStop](tag, "favourite_stops") {
+
+    def id = column[Long]("id", O.PrimaryKey, O.AutoInc)
+    def userId = column[Long]("userId")
+    def stopId = column[String]("stopId")
+
+    def * = (id, userId, stopId) <> ((FavouriteStop.apply _).tupled, FavouriteStop.unapply)
+
+    def userForeignKey = foreignKey("FK_USERS", userId, usersQuery)(
```
Comment (PlamenNeshkov) [inline app/db/UsersRepo.scala]: Adhere to the existing style - `users_fk`

File: app/db/UsersRepo.scala [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  protected val usersQuery = TableQuery[UserTable]
+}
+
+private[db] trait FavouriteStopTable extends UserTable with StopTable { this: SlickComponent =>
+  import dbConfig.profile.api._
+
+  private[FavouriteStopTable] class FavouriteStopTable(tag: Tag)
+    extends Table[FavouriteStop](tag, "favourite_stops") {
```
Comment (PlamenNeshkov) [inline app/db/UsersRepo.scala]: No need for this newline

File: app/db/UsersRepo.scala [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    def userId = column[Long]("userId")
+    def stopId = column[String]("stopId")
+
+    def * = (id, userId, stopId) <> ((FavouriteStop.apply _).tupled, FavouriteStop.unapply)
+
+    def userForeignKey = foreignKey("FK_USERS", userId, usersQuery)(
+      user => user.id, onDelete = ForeignKeyAction.Cascade)
+
+    def stopForeignKey = foreignKey("FK_STOPS", stopId, stopQ)(
+      stop => stop.id)
```
Comment (PlamenNeshkov) [inline app/db/UsersRepo.scala]: No need for this newline

File: app/db/UsersRepo.scala [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    def id = column[Long]("id", O.PrimaryKey, O.AutoInc)
+    def userId = column[Long]("userId")
+    def stopId = column[String]("stopId")
+
+    def * = (id, userId, stopId) <> ((FavouriteStop.apply _).tupled, FavouriteStop.unapply)
+
+    def userForeignKey = foreignKey("FK_USERS", userId, usersQuery)(
+      user => user.id, onDelete = ForeignKeyAction.Cascade)
+
+    def stopForeignKey = foreignKey("FK_STOPS", stopId, stopQ)(
```
Comment (PlamenNeshkov) [inline app/db/UsersRepo.scala]: Adhere to the fk naming style

File: app/db/UsersRepo.scala [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    extends Table[FavouriteStop](tag, "favourite_stops") {
+
+    def id = column[Long]("id", O.PrimaryKey, O.AutoInc)
+    def userId = column[Long]("userId")
+    def stopId = column[String]("stopId")
+
+    def * = (id, userId, stopId) <> ((FavouriteStop.apply _).tupled, FavouriteStop.unapply)
+
+    def userForeignKey = foreignKey("FK_USERS", userId, usersQuery)(
+      user => user.id, onDelete = ForeignKeyAction.Cascade)
```
Comment (PlamenNeshkov) [inline app/db/UsersRepo.scala]: No need for this newline

File: app/db/UsersRepo.scala [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+  def list(): Future[Seq[User]] = db.run {
+    usersQuery.result
+  }
+
+  def retrieve(id: Long): Future[Option[User]] = db.run {
+    usersQuery.filter(_.id === id).result.headOption
+  }
+
+  def addFavouriteStop(userId: Long, stopId: String): Future[FavouriteStop] = db.run {
```
Comment (PlamenNeshkov) [inline app/db/UsersRepo.scala]: This should be in a `FavouriteStopsRepo.create` method

File: app/controllers/UsersController.scala [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+    userRequest match {
+      case Some(request) => {
+        usersRepo.create(request.name, request.password)
+            .map(u => okJson(api.CreateUserResponse.Created(api.UserResponse.fromDb(u))))
+      }
+      case None => {
+        val future = Future { api.CreateUserResponse.NotCreated("Invalid arguments supplied") }
```
Comment (PlamenNeshkov) [inline app/controllers/UsersController.scala]: `NotCreated` should indicate failure to create, assuming the request is valid, eg. database failure

File: app/controllers/UsersController.scala [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+import javax.inject.Inject
+
+import play.api.libs.json.{Json, Writes}
+import play.api.mvc.{AbstractController, Action, ControllerComponents}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class UsersController @Inject()(cc: ControllerComponents, usersRepo: db.UsersRepo)(implicit ec: ExecutionContext)
+  extends AbstractController(cc) {
```
Comment (PlamenNeshkov) [inline app/controllers/UsersController.scala]: Adhere to the indentation/newline existing style for controllers

File: app/db/UsersRepo.scala (hunk 15 above)
Comment (arnaudoff) [reply app/db/UsersRepo.scala]: I was going to do it that way, however, if we have a separate repository for each `User`-`T` relationship for some arbitrary `T` where the relationship is many-to-many, wouldn't this scatter too much boilerplate classes over the project? Moreover, if we have a separate repository for each `T`-`U` many-to-many relationship, this could be rather cumbersome to maintain. Maybe we should implement a separate layer in front of the DAO layer?

File: app/utils/Hasher.scala [hunk 18]
```diff
@@ -0,0 +1,13 @@
+package utils
+
+import org.mindrot.jbcrypt.BCrypt
+
+object Hasher {
```
Comment (PlamenNeshkov) [inline app/utils/Hasher.scala]: Can you extract an interface for this object?

File: app/db/UsersRepo.scala [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+
+protected[db] trait FavouriteStopTable extends UserTable with StopTable { this: SlickComponent =>
+  import dbConfig.profile.api._
+
+  private[FavouriteStopTable] class FavouriteStopTable(tag: Tag) extends Table[FavouriteStop](tag, "favourite_stops") {
+    def id = column[Long]("id", O.PrimaryKey, O.AutoInc)
+    def userId = column[Long]("userId")
+    def stopId = column[String]("stopId")
+    def * = (id, userId, stopId) <> ((FavouriteStop.apply _).tupled, FavouriteStop.unapply)
+    def users_fk = foreignKey("FK_USERS", userId, usersQuery)(user => user.id, onDelete = ForeignKeyAction.Cascade)
```
Comment (PlamenNeshkov) [inline app/db/UsersRepo.scala]: usersFk

File: app/db/UsersRepo.scala [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    def stops_fk = foreignKey("FK_STOPS", stopId, stopQ)(stop => stop.id)
```
Comment (PlamenNeshkov) [inline app/db/UsersRepo.scala]: stopsFk

File: app/controllers/UsersController.scala [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+        usersRepo.create(userCreateRequest.name, userCreateRequest.password)
+          .map(user => okJson(api.UserCreateResponse.Created(api.UserRetrieveResponse.fromDb(user))))
+      }
+    }.getOrElse(Future.successful(BadRequest("Invalid arguments supplied")))
+  }
+
+  def stops(userId: Long, stopId: String) =
+    Authentication(usersRepo.authenticate) {
+      Action.async {
+          usersRepo.addFavouriteStop(userId, stopId)
```
Comment (PlamenNeshkov) [inline app/controllers/UsersController.scala]: Why are we adding a stop in a method called `stops`

Comment (PlamenNeshkov) [reply app/controllers/UsersController.scala]: We would also need a way to retrieve a user's favourite stops

File: app/custom/Authentication.scala [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+import akka.util.ByteString
+import db.User
+import org.apache.commons.codec.binary.Base64.decodeBase64
+import play.api.http.HttpEntity
+import play.api.mvc._
+
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext.Implicits.global
+
+case class Authentication[T](authenticate: (String, String) => Future[Option[User]])(action: Action[T]) extends Action[T] {
```
Comment (PlamenNeshkov) [inline app/custom/Authentication.scala]: We should call this `Authenticated` for better readability

Comment (PlamenNeshkov) [inline app/custom/Authentication.scala]: We should type alias the Strings for username and password for type safety

File: app/custom/Authentication.scala [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+          case Some(user) => action(request)
+          case None => makeResult("Invalid username or password")
+        }
+      }
+      case None => makeResult("You must supply the Authorization header in order to access this page")
+    }
+  }
+
+  def makeResult(error: String): Future[Result] = {
+    val result = Result(
```
Comment (PlamenNeshkov) [inline app/custom/Authentication.scala]: This variable extraction is noisy, you can inline the result in `Future.succesful`

File: app/db/UsersRepo.scala (hunk 19 above)
Comment (PlamenNeshkov) [reply app/db/UsersRepo.scala]: Also, no need for the all caps

File: app/custom/Authentication.scala [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+        authenticate(username, password).flatMap {
+          case Some(user) => action(request)
+          case None => makeResult("Invalid username or password")
+        }
+      }
+      case None => makeResult("You must supply the Authorization header in order to access this page")
+    }
+  }
+
+  def makeResult(error: String): Future[Result] = {
```
Comment (PlamenNeshkov) [inline app/custom/Authentication.scala]: `makeError` is more appropriate with an argument name `message`

File: app/custom/Authentication.scala [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+    val result = Result(
+      header = ResponseHeader(401, Map.empty),
+      body = HttpEntity.Strict(ByteString(error), Some("text/plain"))
```
Comment (PlamenNeshkov) [inline app/custom/Authentication.scala]: It's better to return a json structure instead. We can define an error class for this

File: app/custom/Authentication.scala (hunk 22 above)
Comment (PlamenNeshkov) [inline app/custom/Authentication.scala]: We use the same authentication function for now, we could make it a default value

File: app/controllers/UsersController.scala [hunk 26]
```diff
@@ -2,14 +2,14 @@ package controllers
...
-import custom.Authentication
+import custom.Authenticated
 import play.api.libs.json.{Json, Writes}
 import play.api.mvc.{AbstractController, ControllerComponents}
 
 import scala.concurrent.{ExecutionContext, Future}
 
 class UsersController @Inject() (
-  cc: ControllerComponents, usersRepo: db.UsersRepo,
+  cc: ControllerComponents, implicit val usersRepo: db.UsersRepo,
```
Comment (PlamenNeshkov) [inline app/controllers/UsersController.scala]: Why `implicit val`?

File: app/controllers/UsersController.scala [hunk 27]
```diff
@@ -32,13 +32,14 @@ class UsersController @Inject() (
       }
     }
 
+  def listFavouriteStops(userId: Long) = ???
```
Comment (PlamenNeshkov) [inline app/controllers/UsersController.scala]: lol
